src/main/scala/org/constellation/snapshot/SnapshotSelector.scala

```diff
@@ -1,34 +1,34 @@
 package org.constellation.snapshot
 
+import cats.effect.{Concurrent, ContextShift}
+import cats.implicits._
 import org.constellation.domain.schema.Id
-import org.constellation.storage.{RecentSnapshot, SnapshotVerification}
+import org.constellation.p2p.PeerData
+import org.constellation.storage.SnapshotVerification
 import org.constellation.util.SnapshotDiff
 
-case class DownloadInfo(diff: SnapshotDiff, recentStateToSet: List[RecentSnapshot])
+abstract class SnapshotSelector[F[_]: Concurrent, T <: AnyRef] {
 
-trait SnapshotSelector {
+  type SnapshotSelectionInput = T
 
-  type NodeSnapshots = (Id, List[RecentSnapshot])
+  type NodeSnapshots = (Id, SnapshotSelectionInput)
+  type SelectionInfo = (SnapshotDiff, SnapshotSelectionInput)
 
   def selectSnapshotFromRecent(
     peersSnapshots: List[NodeSnapshots],
-    ownSnapshots: List[RecentSnapshot]
-  ): Option[DownloadInfo]
+    ownSnapshots: SnapshotSelectionInput
+  ): Option[SelectionInfo]
 
   def selectSnapshotFromBroadcastResponses(
     responses: List[Option[SnapshotVerification]],
-    ownSnapshots: List[RecentSnapshot]
-  ): Option[DownloadInfo]
-
-  private[snapshot] def createDiff(
-    major: List[RecentSnapshot],
-    ownSnapshots: List[RecentSnapshot],
-    peers: List[Id]
-  ): SnapshotDiff =
-    SnapshotDiff(
-      ownSnapshots.diff(major).sortBy(_.height).reverse,
-      major.diff(ownSnapshots).sortBy(_.height).reverse,
-      peers
+    ownSnapshots: SnapshotSelectionInput
+  ): Option[SelectionInfo]
+
+  def collectSnapshot(
+    peers: Map[Id, PeerData]
+  )(cs: ContextShift[F])(implicit m: Manifest[SnapshotSelectionInput]): F[List[(Id, SnapshotSelectionInput)]] =
```

Comment:
is the `Id` unique? Maybe it can be a `Map[Id, SnapshotSelectionInput]` instead of `List[(Id, SnapshotSelectionInput)]`